### ISSUE TITLE
Move the subgraph label to the bottom left corner

### ISSUE THREAD

**Issue by Disshell** [mermaid-js/mermaid#2977]
I have the following graph:
```
graph TD

    A(create_bot.py) -->|dp| B(main_bot.py)
    subgraph handlers.py
        C(client.py) --- F
        D(admin.py) --- F
        E(other.py) --- F
        F[handlers]
    end
    A --> |dp, bot| C
    A --> |dp, bot| D
    A --> |dp, bot| E
    F --> |client, admin, other| B
    style handlers.py text-align:left
```
i want handlers.py to be in the bottom left corner of the square.
![uZ9e2gIV8I](https://user-images.githubusercontent.com/23436995/165566474-c1f03b6d-59bb-4a10-9fd4-2e2235f8d02d.png)

**Comment by flywire** [comment]
No, need more flexible label alignment in general.
```
flowchart TD
 A-->B
 subgraph gb
  B
 end
```
![image](https://user-images.githubusercontent.com/11288701/172099808-3d994b3f-9baf-42ef-9aa0-8a421c4824e4.png)

**Comment by flywire** [comment]
Useful alignments.

1. Horizontal:
   * left
   * centre
   * right
   * indent
   * undent

1. Vertical:
   * top
   * middle
   * bottom

**Comment by knsv** [comment]
Good idea. Pu the label in a suitable position. Adding to the backlog.

**Comment by sjackson0109** [comment]
Same question for ICONS/images:

```
flowchart LR
    subgraph Border
        subgraph Hub[HUB]
            direction TB
            subgraph Subnet1["Subnet Name <br /> 10.1.1.0/24"]
                Subnet1Icon["<br /><img src='../images/networking/subnet.svg' />..."]
            end
            vNetIcon["<br /><img src='../images/networking/vnet.svg' />..."]
        end
    end
%% Defining Class Styles
%% classDef <Name> COLOR:TEXTCOLUR, FILL:BGFILLCOLOR,STROKE:BORDERCOLOR,STROKE-WIDTH:SIZEOFBORDER,STROKE_DASHARRAY:LINE GAP;
classDef vNet fill:#fff,stroke:#4698eb,stroke-width:2px,color:#000,stroke-dasharray: 2 4;

classDef Subnet fill:#dfe5f3,stroke:#787675,stroke-width:1px,color:#000,stroke-dasharray: 2 4;

classDef Icon margin:0px, stroke-width:0px, padding:0px, fill:transparent, position:absolute, bottom:0px, right:0px;

%% Custom Styles
style Border fill:#fff,stroke:#fff,stroke-width:4px,color:#fff,stroke-dasharray: 1 0;

%% Assigning Nodes to Classes
class Border Border;
class Hub vNet;
class Subnet1 Subnet;
class Subnet1Icon,vNetIcon Icon;
```

This is what it looks like:
![image](https://user-images.githubusercontent.com/38080190/200002023-b128eb53-5aeb-45a5-b3f2-ec979d5fd0a7.png)

don't really mind which corner it anchors too, but CSS would normally be used, like so:
- Style='position:absolute,bottom:10px, right:10px;'


What is the mermaid syntax for this.. i've spent hours trying and cannot figure it out.
As you can see i've tried:
classDef Icon margin:0px, stroke-width:0px, padding:0px, fill:transparent, `position:absolute, bottom:0px, right:0px;`
Which does change the alignment, but only within the Subgraph's original shape/position. so still 'position:relative'.

**Comment by flywire** [comment]
> i've spent hours trying and cannot figure it out.

Intuitively this type of manipulation seems easy but impossible in practice. The user guide should give more detail/examples of what _is_ possible.

**Comment by sjackson0109** [comment]
For those of you who can't wait to get an answer...
I did actually manage to align it, but it's not pretty...
![image](https://user-images.githubusercontent.com/38080190/201249070-afc3c243-79e6-4b45-923a-849c5870286d.png)

```
flowchart LR
         subgraph IDENTITY["Identity <br /> 10.10.10.0/24"]
            direction TB
            subgraph 14[ ]
                direction LR
                IDSubnet1("Domain Controllers <br /> 10.10.10.0/29")
                SPACER20PX
            end
            IDvNetIcon["<br /><br /><img class='Icon' src='./img/vnet.svg' style='bottom:0px;right:0px;' />..."]
        end
classDef vNet fill:#dfe5f3,stroke:#4698eb,stroke-width:2px,color:#000,stroke-dasharray: 8 4;
classDef Subnet fill:#fff,stroke:#4698eb,stroke-width:1px,color:#000,stroke-dasharray: 4 8;
classDef Icon margin:0px, border:0px,stroke-width:0px, padding:0px, color:transparent, fill:transparent, position:absolute,right:0px, bottom:0px;

class IDENTITY vNet;

style IDENTITY height:160px,width:325px;

style SPACER20PX fill:transparent,stroke:transparent,color:transparent,height:10px;

```

The use of `<br />` gave me the vertical height, width attribute on the Icon classDef worked nicely too; and i changed the alignment with `position:absolute, right:0, bottom:0;`.  

Don't ask me why i have to add 3x dots `...` to the end of the IMG definition, the image just won't render without them!

**Comment by flywire** [comment]
Your example does not render in the [Live Editor](https://mermaid.live/edit#pako:eNqFVF1v2yAU_SuIPGSTPH8lcRN3mbQ1fYhUVVWTh031HrABGw2DBThtFOW_D38kjZdpQ5aFz72cezhcfICZxATGkHL5mhVIGfDwnAhwGrpOc4WqAqxX94_b9fbHSwLXmAjDzB58ThXwvoDAd7vH98JpAn9eLLcDM0Uyw6QA22_DyJk7mL6AP1YNVw4k9WO92tSpICb4kMCVLBET4E4KoyTnROm_aVsk8OM1z-bp6939c-g_fR_GiMBDYL3aPRKzzqSwHnT0_ZuVOcg40no5bsJjoFW2HLuexb2dlejqXW5Bs-dkOU6lMbKM_ertVrG8MO1sbGlc1x2Y1wpoaVeEgqY2oIzzeIQpmdGJo-1ef5F4NI0Wc5L2n59eGTZFHFZvTia5VPHI9_1TDCNtT1ihfQzmYHp7wd5Z2fNTSv9JHvyPfArml-SNJ6BEKmei2awDUqkwUe18QNwGK4QxE3n30ZUxCgldIWXbzuk0DpBKata0SYxSLXltiHM2tql1tjsRvahzM7eutnh7OO94QVqGIGo4OnGTcNZzdLnvfXMtqXfvErreyalGJw06sCTKtjG2l_HQdEECTUFKksDYTjGhqOYmgYk42tS6wsiQe8yMVDCmiGviQFQbudmLDNoyNTklrRiyt6w8Z3GJrPkwPkCzr5qbnzNtLKU9JMryBq8Vt3BhTKVjz2vCbs5MUaduJktPM9z8JordIvKiMJqjcEKimwmaTSY4S4PFnIbTgOIbPwgRPB6PvwGwRmYb).

**Comment by sjackson0109** [comment]
@flywire you are funny :)

(Of course you need a public facing URL inside the IMG tag, to an appropriate SVG image... try this) 
`https://symbols.getvecta.com/stencil_28/71_virtual-network.8cd684329b.svg`


```
flowchart LR
         subgraph IDENTITY["Identity <br /> 10.10.10.0/24"]
            direction TB
            subgraph 14[ ]
                direction LR
                IDSubnet1("Domain Controllers <br /> 10.10.10.0/29")
                SPACER20PX
            end
            IDvNetIcon["<br /><br /><img class='Icon' src='https://symbols.getvecta.com/stencil_28/71_virtual-network.8cd684329b.svg' style='bottom:0px;right:0px;' />..."]
        end
classDef vNet fill:#dfe5f3,stroke:#4698eb,stroke-width:2px,color:#000,stroke-dasharray: 8 4;
classDef Subnet fill:#fff,stroke:#4698eb,stroke-width:1px,color:#000,stroke-dasharray: 4 8;
classDef Icon margin:0px, border:0px,stroke-width:0px, padding:0px, color:transparent, fill:transparent, position:absolute,right:0px, bottom:0px;

class IDENTITY vNet;

style IDENTITY height:160px,width:325px;
style 14 stroke-width:0px, padding:0px, fill:transparent,color:transparent,width,130px,margin:0px,padding:0px,lineHeight:0px;
style SPACER20PX fill:transparent,stroke:transparent,color:transparent,height:10px;
```

Clearly live-editor doesn't do a good job of applying the styles/classes properly, especially the iconDef class, with colour and fill transparency:
![image](https://user-images.githubusercontent.com/38080190/201304629-e06c4378-3b4a-4f94-a513-b10bfd8879a7.png)

The same MD file in Visual Studio Code, with Mermaid Preview:
![image](https://user-images.githubusercontent.com/38080190/201304580-69da4856-59de-4057-9fcc-1301594be37c.png)

**Comment by flywire** [comment]
> you need a public-facing URL inside the IMG tag, to an appropriate SVG image

My concern was the shape of the boxes and I see `style 14` fixed that, and surprises me that the first example was generated.

**Comment by sjackson0109** [comment]
@flywire - creating a cascade of nested shapes to ‘push the icon into position’ is NOT a solution, it’s a workaround.

draw a box, put an image inside, set position to absolute, then bottom-right align it.  It works in CSS , outside of mermaid, with a div and image. But not using mermaid shapes.

In-fact mermaid shapes with a very small number of characters in the body also don’t render images properly neither. Notice I had 3x full-stops, and 2x line breaks in my ‘IDvNetIcon’ shape?

**Comment by underscoremin** [comment]
I am also experiencing issues with label positioning:

flowchart TD
    subgraph Root
        n1(.)
    end
    subgraph TLD [Top Level Domains]
        n1 --> n2(.com)
        n1 --> n4(.org)
        n1 --> n6(...)
    end
    subgraph SLD [Second Level Domains]
        n2 --> n7(google.com)
        n2 --> n9(...)
        n4 --> n8(wikipedia.org)
        n4 --> n10(...)
    end

![Screenshot 2022-12-18 at 9 29 10 am](https://user-images.githubusercontent.com/118170565/208268259-d5935b5c-939f-4861-93e4-4157134ffd6f.png)

**Comment by flywire** [comment]
Hey, that's crazy in dark mode. Testing `<br>` (click image for code):

[![image](https://user-images.githubusercontent.com/11288701/208279323-2fdeb125-bc20-41f2-a25f-53ee43607ee2.png)](https://mermaid.live/edit#pako:eNplUUFuwjAQ_IrlU5AgIQkNEFWc0hunwq3uwYk3jlXHjhwbhFD-XrtAi9TbemZ2Znd9xY1mgEvcSn1uOmosOlZEja7mhg4detfaEqXSKJ4RBYo9Ucd9hT6OekB7OIFEle6pUONnUKPFYodUFsWN7md_wCqKteFPQBHF8X_jQzA-QKMVe63N7sc-FE8J2a1_HXGtuYRHzh3e3m3V6vbeRGfxJQZggj4GuDPp8m8CPMc9GB_B_DmuRCFEsO2gB4JLXzJoqZOWYKImL3UDoxbemLDa4LKlcoQ5ps7qw0U1uLTGwUNUCeoX639VUlMGvumK7WUIt-ditN7SL9wKHnBnpIc7a4exTJJAx1zYztVh0WQULHxUd9oWSZEVG5rlUKxz-pLnrKnT7abNVmnL1ss0o3iapm-MKp0q)

**Comment by underscoremin** [comment]
Yeah using ```<br>``` doesn't seem to help unfortunately.

**Comment by CrBatista** [comment]
Definitely, not the best solution, but a workaround to align left or right by adding spaces to the subgraph title:
```
flowchart TD
    A[System A] --> B[System B]
    subgraph Title&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp&nbsp
        B -->|No| C(System D)
        B -->|Yes| D(External System)
    end
```


![github](https://user-images.githubusercontent.com/17290362/224781133-96e67170-37f8-4c40-86a9-50431880fbd1.png)

**Comment by verygoodsoftwarenotvirus** [comment]
Just voicing my support/desire for this feature. I actually would love the abillity to have an upper and lower label, if possible.

**Comment by clarkezone** [comment]
+1 to this feature request

**Comment by ryanjoneil** [comment]
In case you need another motivating example, I'd like to use Mermaid to draw decision diagrams top-down. [Here's an example](https://ryanjoneil.github.io/posts/2023-09-13-visualizing-decision-diagrams/#mermaid).

Aside: it would be nice if it were possible to align subgraph edges, too.

**Comment by jeroenheijmans** [comment]
Adding a somewhat more minimal repro for a typical scenario where the subgraph label is nog very usable:

```
flowchart TD
    Root ==> Child
    subgraph Items
    Child --> ThingA
    Child --> ThingB
    Child --> ThingC
    end
```

Which renders with the fat `Root ==> Child` arrow on top of the subgraph label:

![sample graph](https://github.com/mermaid-js/mermaid/assets/1590536/c24517c0-bb11-436e-9ba0-60e6349a5f52)

Having the nine corner positions as "anchors" would work for most of my use cases.

**Comment by flywire** [comment]
> Having the nine corner positions as "anchors" would work for most of my use cases.

```
LT CT RT
LM CM RM
LB CB RB
```

**Comment by wyatt-troia-msft** [comment]
Bumping this. This would be incredibly helpful. I want to transition my team to using Mermaid but it's hard to justify if labels aren't legible.

**Comment by sjackson0109** [comment]
> > Having the nine corner positions as "anchors" would work for most of my use cases.
> 
> ```
> LT CT RT
> LM CM RM
> LB CB RB
> ```
TOTALLY AGREE

Think it should support a further 9x 
```
TL TC TR
ML MC MR
BL BC BR
```

**Comment by ryeleo** [comment]
> Bumping this. This would be incredibly helpful. I want to transition my team to using Mermaid but it's hard to justify if labels aren't legible without hacky fixes.

This is precisely how I am feeling. Mermaid.js flow charts are **_so very good_**, but I definitely feel like 'subgraph label placement' is a required feature to make it useable for the scenarios my colleagues and I run into.

<img width="509" alt="image" src="https://github.com/mermaid-js/mermaid/assets/2552438/ca386ff2-8b3b-4fd3-8a42-e58295b46e53">

**Comment by zlatko-minev** [comment]
+1 to this feature request

**Comment by sjrothfuss** [comment]
+1 to this feature request

**Comment by cmaster11** [comment]
+1 for this feature 🌻

**Comment by lxwulf** [comment]
1+ for this feature request :1st_place_medal:

**Comment by droumis** [comment]
+1

**Comment by cwoodall-cfs** [comment]
+1

**Comment by raphaelquast** [comment]
+1 for this feature!

Since this is still an open issue and I couldn't find a "suitable workaround" to move labels in the comments... here's mine:

Just wrap the title in a `<div style="margin: ...">my title</div>` and adjust margins as you like in relative `ex` units.  
(best done once the graph is finished to get the margins right since subgraph border will adjust according to the margins!)

<details> <summary>click to view code</summary>

```
graph TD
    A[aaa]
    
    subgraph B["normal label"]
    bb
    bbbb
    end

    subgraph C["<div style="margin:8ex 0 0 12ex;">RIGHT label</div>"]
    cc
    cccc
    end

    subgraph D["<div style="margin:8ex 52ex 0 0;">LEFT label</div>"]
    dd
    ddd
    dddd
    end

    subgraph E["<div style="margin:8ex 0 0 0;">bottom label</div>"]
    ee
    eee
    eeee
    end

    A-.-> C
    C-.-> D
    B-.-> E

    A -->  cccc
    cccc --> dd
    bb --> ee & eee & cc & dd
    bbb --> dd
    bbbb --> ddd
```

</details>

```mermaid
flowchart TD
    A[aaa]
    
    subgraph B["normal label"]
    bb
    bbbb
    end

    subgraph C["<div style="margin:7.5ex 0 0 12ex;">RIGHT label</div>"]
    cc
    cccc
    end

    subgraph D["<div style="margin:7.5ex 52ex 0 0;">LEFT label</div>"]
    dd
    ddd
    dddd
    end

    subgraph E["<div style="margin:7.5ex 0 0 0;">BOTTOM label</div>"]
    ee
    eee
    eeee
    end

    A-.-> C
    C-.-> D
    B-.-> E

    A -->  cccc
    cccc --> dd
    bb --> ee & eee & cc & dd
    bbb --> dd
    bbbb --> ddd
```

**Comment by Callum-Halton** [comment]
+1

**Comment by flywire** [comment]
> _suitable workaround_

@raphaelquast Can you adjust the block label spacing so the labels can be clearly seen? If so this workaround should be in the docs.

**Comment by sjackson0109** [comment]
I would prefer to write my emails in binary equivalent of each ascii character, sooner than waste my time padding and spacing and carriage returning a html block, to fit a string/image inside a mermaid subgraph label looking nice.What you suggest is a MASSIVE waste of time. Nobody would do it more than once. Please don’t consider publishing that to the world..On 5 Sep 2024, at 10:54, flywire ***@***.***> wrote:﻿

suitable workaround

@raphaelquast Can you adjust the block label spacing so the labels can be clearly seen? If so this workaround should be in the docs.

**Comment by anomyq** [comment]
+1

**Comment by maneesh29s** [comment]
Any updates on this feature?

**Comment by AdmilsonGoncalves** [comment]
+1 for this request

**Comment by gaodi16366** [comment]
+1 for this request

**Comment by NatalieZelenka** [comment]
+1 for this request

**Comment by vincentor** [comment]
+1 for the request

**Comment by yoanhuret** [comment]
+1 too !

**Comment by gojack10** [comment]
+1

**Comment by peter-mills-dev** [comment]
+1

**Comment by mrchantey** [comment]
This can be done with css flexboxes.

```
graph TD
  subgraph F ["<div style="width:30em; height:8em; display:flex; justify-content: flex-start; align-items:flex-end;">Title</div>"]
    foo
    bar
  end
```

```mermaid
graph TD
  subgraph F ["<div style="width:30em; height:8em; display:flex; justify-content: flex-start; align-items:flex-end;">Title</div>"]
    foo
    bar
  end
```

**Comment by sjackson0109** [comment]
I don't really see what you are demonstrating here @mrchantey 

![Image](https://github.com/user-attachments/assets/2c9c0c99-6d50-4e46-82cb-635975e00437)

**Comment by flywire** [comment]
https://css-tricks.com/snippets/css/a-guide-to-flexbox/

**Comment by flywire** [comment]
```mermaid
graph TD
subgraph F ["<div style="width:30em; height:8em; display:flex; justify-content: center;">Title</div>"]
foo
bar
end
```

**Comment by koenklomps** [comment]
+1

**Comment by lushkovsky-s** [comment]
+1

For me, it ruins 50% of subgraph usages, would be extremely greateful for fast track for this one

**Comment by otac0n** [comment]
This would help with mitigating overlapping labels as well:

<img width="2053" height="747" alt="Image" src="https://github.com/user-attachments/assets/bad1d6eb-829b-44d9-943f-6b7b7b801fa8" />

**Comment by knsv** [comment]
Yeah, let's make it configurable and perhaps some auto option as well, so that if the subgraph label is occupied, it moves to some place where it's free. 

You could see that you want to put it on the left and right sides, maybe rotated as well.

**Comment by knsv** [comment]
<html><head></head><body>
<h1>Suggested syntax and first implementation</h1>
<p>The title lives on one of four edges. The value is <code>edge</code> or <code>edge-anchor</code>. The first token is always the edge, and it fixes the orientation: top/bottom run horizontal, left/right run rotated 90°. The second token is where the title sits along that edge. A bare edge name means centered, consistent with <code>top</code> meaning top-center today.</p>
<p>This renames nothing: the current six values are exactly the top and bottom rows of a 12-cell grid.</p>

edge | orientation | values (near end : center : far end)
-- | -- | --
top | horizontal | top-left : top : top-right
bottom | horizontal | bottom-left : bottom : bottom-right
left | vertical, reads bottom→top | left-top : left : left-bottom
right | vertical, reads top→bottom | right-top : right : right-bottom


<pre><code>               top-left         top         top-right
              ┌────────────────────────────────────┐
     left-top │                                    │ right-top
              │                                    │
        left  │                                    │ right
              │                                    │
  left-bottom │                                    │ right-bottom
              └────────────────────────────────────┘
             bottom-left       bottom       bottom-right
</code></pre>
<p><strong>Must handle:</strong> <code>top-left</code> and <code>left-top</code> are distinct and both valid. <code>top-left</code> is on the top edge at its left end (horizontal text). <code>left-top</code> is on the left edge at its top end (vertical text). Both hug the same corner with different orientation. Docs should show that pair side by side.</p>
<p><strong>Rotation defaults per edge:</strong> left reads bottom→top (90° counter-clockwise), right reads top→bottom (90° clockwise), so both labels lean outward and read balanced around the box. Expose an override as its own field rather than encoding it in the position:</p>
<pre><code>title-position left-top
title-rotation ccw        # auto | cw | ccw, default auto
</code></pre>
<p> I worry a little about nodes called "title-position". This change might render those diagrams with syntax errors. An alternative is to only keep this parameter available using the @ syntax.</p><p>Metadata form we converge on later:</p>
<pre><code>subgraph id [Title]@{ titlePosition: left-top, titleRotation: ccw }
</code></pre>
<p><code>auto</code> resolves to the per-edge defaults above. Keeping rotation separate means <code>titlePosition</code> stays a single token for effectively all real usage.</p>
<p><strong>Leave the door open for RTL (do not build now):</strong> the anchor tokens should be able to accept <code>start</code> / <code>end</code> as logical aliases later (left/right on horizontal edges, top/bottom on vertical) without changing the concrete names. Just don't design anything that blocks it.</p>
<p>Net: <code>titlePosition</code> carries edge plus anchor across all 12 cells, <code>titleRotation</code> is the rare override, and none of the current six values change.</p><p>I will make a default, auto mode that starts in this order: top, top-left, bottom, bottom-left, top-right, bottom-right. If everything fails, then use top.</p></body></html>
